### ISSUE TITLE
Multigrid enhancements

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#2328]
@kronbichler -- I have Karl Ljungkvist (@kalj) in my office and we are discussing multigrid in the context of matrix free operations. He says you talked about extensions in this direction. Could you summarize what you have in mind so we can coordinate?

**Comment by guidokanschat** [comment]
Since Martin is not available right now: he already implemented [MGTransferMatrixFree](https://www.dealii.org/8.4.0/doxygen/deal.II/classMGTransferMatrixFree.html). People in Heidelberg are working on matrix free smoothers.

**Comment by kalj** [comment]
Thanks. So, the motivation is that prolongation and restriction was done using matrices previously? In that case, I guess it has similar performance characteristics as the operator application and thus, should also benefit from being matrix free. Am I right?

**Comment by guidokanschat** [comment]
Yes. Prolongation and restriction have very regular structures and the matrices only depend on the finite elements chosen, not even on the operator. Therefore, it is even more obvious that they should be implemented matrix free.

**Comment by kalj** [comment]
Sounds good. About the smoothing, in my limited understanding, isn't that already matrix free with a MGSmootherPrecondition configured with a matrix-free operator? (I'm thinking of Step-37 now, with the Chebyshev preconditioner)

**Comment by guidokanschat** [comment]
That is true, but we are looking into efficient domain decomposition smoothers

**Comment by kronbichler** [comment]
I think that the basic matrix-free transfer operations are in place. So I think that you @kalj can start looking into how MGTransferMatrixFree is organized for a variant on the GPU code.

I also have a modification of the main multigrid algorithm in the pipeline which does the transfer of the defects differently to the main Multigrid class which I will prepare for merging at some point during the next weeks (I realized that the old algorithm does too many global communication steps which hurts on 50,000+ cores: on 1k cores you can scale strongly down to around 20k dofs per rank or 0.015s per V-cycle, but on 100k cores you scale only down to 100k dofs per core or 0.05s per V-cycle). This includes some additional options for the transfer classes.

Another thing that is planned from my side is to support systems of PDEs with MGTransferMatrixFree similarly to some MGTransferSelect options, but that is further away.

**Comment by kalj** [comment]
Great, I will take a look at it during the coming weeks.

**Comment by kronbichler** [comment]
Superseded by #2534.
